### PR TITLE
`ledger-api-bench-tool` - support for multiple streams [DPP-372]

### PR DESCRIPTION
**Pull request by kamil-da**
### Main changes
- support for reading multiple streams
- configuration parameters for the gRPC thread pool
- actor-based metric handling

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala [hunk 1]
```diff
@@ -16,47 +23,67 @@ import java.util.concurrent.{
...
-        val ledgerIdentityService: LedgerIdentityService = new LedgerIdentityService(channel)
-        val ledgerId: String = ledgerIdentityService.fetchLedgerId()
-        val transactionService = new TransactionService(channel, ledgerId)
-        config.streamConfig.foreach { streamConfig =>
+    resources.use { case (channel, system) =>
+      val ledgerIdentityService: LedgerIdentityService = new LedgerIdentityService(channel)
+      val ledgerId: String = ledgerIdentityService.fetchLedgerId()
+      val transactionService = new TransactionService(channel, ledgerId)
+      Future
+        .traverse(config.streams) { streamConfig =>
```
Comment (tudor-da) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala:58]: The parallelism in this traverse is theoretically bounded by the ExecutionContext size, and could lead to confusing results if more streams than CPU cores are started

Comment (kamil-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala:58]: Please correct me if I'm wrong, but having a separately configured thread pool for the gRPC channel (line 48) should allow us to run multiple transaction streams with requested concurrency parameters. The global execution context will be used just for spinning up all those streams.

Comment (tudor-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala:58]: Oh, those services are using the configured thread pool.. Sometimes it's just better to review in the IDE! Apologies

Comment (tudor-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/LedgerApiBenchTool.scala:58]: Oh, those services are using the configured thread pool.. Sometimes it's just better to review in the IDE! Apologies

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala [hunk 2]
```diff
@@ -26,30 +30,49 @@ object Cli {
...
+        )
+        .action { case (streamConfig, config) =>
+          config.copy(streams = config.streams :+ streamConfig)
+        },
+      opt[FiniteDuration]("log-interval")
+        .abbr("r")
+        .text("Stream metrics log interval.")
+        .action { case (period, config) => config.copy(reportingPeriod = period) },
+      opt[Int]("core-pool-size")
+        .text("corePoolSize parameter of the ThreadPoolExecutor.")
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala:49]: users are not interested in such an implementation detail
```suggestion
        .text("Size of the worker thread pool.")
```

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala [hunk 3]
```diff
@@ -26,30 +30,49 @@ object Cli {
...
+        .optional()
+        .action { case (size, config) =>
+          config.copy(concurrency = config.concurrency.copy(corePoolSize = size))
+        },
+      opt[Int]("max-pool-size")
+        .text("maxPoolSize parameter of the ThreadPoolExecutor.")
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala:55]:
```suggestion
        .text("Maximum size of the worker thread pool.")
```

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala [hunk 4]
```diff
@@ -26,30 +30,49 @@ object Cli {
...
+        .optional()
+        .action { case (size, config) =>
+          config.copy(concurrency = config.concurrency.copy(maxPoolSize = size))
+        },
+      opt[Long]("keep-alive-time")
+        .text("keepAliveTime parameter of the ThreadPoolExecutor.")
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala:61]: Do we really need to expose such a parameter? It is unlikely anyone will ever need to change it. It makes CLI less transparent

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala [hunk 5]
```diff
@@ -26,30 +30,49 @@ object Cli {
...
+        .optional()
+        .action { case (time, config) =>
+          config.copy(concurrency = config.concurrency.copy(keepAliveTime = time))
+        },
+      opt[Int]("max-queue-length")
+        .text("maxQueueLength parameter of the ThreadPoolExecutor.")
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala:67]: As above

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/MetricalStreamObserver.scala [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.benchtool.metrics
+
+import akka.actor.typed.ActorRef
+import org.slf4j.Logger
+
+class MetricalStreamObserver[T](
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/MetricalStreamObserver.scala:9]: The word metrical sounds a bit too poetic. Elsewhere we use the word Metered or Timed in similar contexts.
```suggestion
class MeteredStreamObserver[T](
```

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/ObserverWithResult.scala [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    ()
+  }
+
+  override def onError(t: Throwable): Unit = {
+    logger.error(withStreamName(s"Received error: $t"))
+    promise.failure(t)
+  }
+
+  override def onCompleted(): Unit = {
+    logger.debug(withStreamName(s"Completed."))
```
Comment (mziolekda) [inline ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/ObserverWithResult.scala:29]: You used to log completion at an info level. I would keep one message at the start and one at the end of a stream with that level.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/MetricalStreamObserver.scala (hunk 6 above)
Comment (kamil-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/MetricalStreamObserver.scala:9]: I would say that the name `MeteredStreamObserver` suggests that the observer itself is metered but I'm not 100% sure :)
If you're convinced to the `MeteredStreamObserver` I will change the name.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala (hunk 4 above)
Comment (kamil-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala:61]: You're right. I'll remove this one.

File: ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala (hunk 5 above)
Comment (kamil-da) [reply ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/Cli.scala:67]: 👍  I'll remove this.

**Comment by kamil-da** [comment]
@mziolekda Thank you for the review! I've just done fixes according to your suggestions and it's ready for the second check.

**Review by mziolekda — APPROVED** [review]
It is very nifty metrics gathering system you have created here. Well done!

**Pull request merged**